### ISSUE TITLE
XML namespace errors

### ISSUE THREAD

**Issue by woutdenolf** [nexusformat/definitions#1031]
I'm using the [XML language plugin by Red Hat](https://marketplace.visualstudio.com/items?itemName=redhat.vscode-xml) in VScode

I'm getting these errors:

![image](https://user-images.githubusercontent.com/7264703/157903322-2d7e9320-3758-4027-b8e7-8ac040fb7395.png)

I'm not an XML expert but the file http://definition.nexusformat.org/nxdl/nxdlTypes.xsd is missing. Perhaps that causes some of the errors?

**Comment by prjemian** [comment]
It's here: https://github.com/nexusformat/definitions/blob/0592a9881d63d165739b6fa554b0d0ecc3ff4b7a/nxdl.xsd#L35-L42

How is this tool trying to find it?  Perhaps assuming the `targetNamespace` URI is a valid URL? https://github.com/nexusformat/definitions/blob/0592a9881d63d165739b6fa554b0d0ecc3ff4b7a/nxdl.xsd#L4

It is not required for the `targetNamespace` to be a valid web address, or even formatted as such, but it is customary for the formatting.  I'm guessing this tool makes the assumption that the `nxdlTypes.xsd` file will be served from this web address.  In our case, `targetNamespace` is a [URI that is not a URL](https://stackoverflow.com/questions/176264/what-is-the-difference-between-a-uri-a-url-and-a-urn).

For reference about `targetNameSpace`, this [SO post](https://stackoverflow.com/questions/9853484/why-do-we-need-targetnamespace#9909529) has helpful explanations.  The W3schools has reference documentation on [`targetNamespace`(https://www.w3schools.com/xml/el_schema.asp) and the related [`xmlns`](https://www.w3schools.com/XML/xml_namespaces.asp).

**Comment by woutdenolf** [comment]
The tools gets the namespace information from http://definition.nexusformat.org/nxdl/nxdl.xsd but the file nxdlTypes.xsd does not exist there...

**Comment by prjemian** [comment]
Note that since `targetNamespace` is not a URL (in our case), then attempts to access a URL the file http://definition.nexusformat.org/nxdl/nxdlTypes.xsd give a 404 (not found) error, as expected.  Nothing to fix here other than the assumption that a URI must always be a working URL.

**Comment by prjemian** [comment]
As discussed in #92 and answered in #835.  Also referenced in #212.

**Comment by woutdenolf** [comment]
So this means I can only work on new nxdl.xml files inside the definitions repo and not anywhere else?

**Comment by woutdenolf** [comment]
The tools gets the namespace from
```xml
xsi:schemaLocation="http://definition.nexusformat.org/nxdl/3.1 ../nxdl.xsd"
```
which I had to change to this to work
```xml
xsi:schemaLocation="http://definition.nexusformat.org/nxdl/3.1 http://definition.nexusformat.org/nxdl/nxdl.xsd"
```
except that it doesn't work because "nxdlTypes.xsd" is not availabe in "http://definition.nexusformat.org/nxdl/".

**Comment by woutdenolf** [comment]
> So this means I can only work on new nxdl.xml files inside the definitions repo and not anywhere else?

Isn't the idea of publishing an xml namespace that it can be used by validation tools everywhere?

**Comment by prjemian** [comment]
> So this means I can only work on new nxdl.xml files inside the definitions repo and not anywhere else?

If you want to use such automation tools that assist you in editing and checking validity of NXDL (.nxdl.xml) files, then the answer is Yes.  Inside a branch or fork.

**Comment by prjemian** [comment]
> Isn't the idea of publishing an xml namespace that it can be used by validation tools everywhere?

Pick up the previous discussion: https://github.com/nexusformat/definitions/issues/835#issuecomment-712943011

Propose the publishing of this URI as an actual URL in a new issue.

**Comment by woutdenolf** [comment]
Having a URL for nxdl.xsd but not for nxdlTypes.xsd makes no sense imo. Either you have a URL for all .xsd files or for none.

However the .xsd files refer to each other with URI's:

https://github.com/nexusformat/definitions/blob/0592a9881d63d165739b6fa554b0d0ecc3ff4b7a/nxdl.xsd#L35-L42

So I'm not sure this is even going to solve the problem.

**Comment by woutdenolf** [comment]
https://github.com/nexusformat/definitions/blob/0592a9881d63d165739b6fa554b0d0ecc3ff4b7a/nxdl.xsd#L3

> xs is the namespace of the XML Schema for XSD

From [this](https://en.wikipedia.org/wiki/XML_namespace#Namespace_declaration) and [this](https://www.w3schools.com/xml/schema_schema.asp) description it seems that `xs` is the _prefix_ of the "http://www.w3.org/2001/XMLSchema" namespace (a.k.a. the _XML Schema_, _XML Schema Definition_ or _XDS_ namespace). The name of an XML namespace is often a URL to avoid global collisions.

>  The re-use of terms here is really mind numbing.

Yes I'm losing my mind over here. For example _XML Schema_ is apparently one of the several [XML schema languages](https://en.wikipedia.org/wiki/XML_schema#Languages). That's like having a car brand called `Car`. Do you have a car, a `Car` or both?

**Comment by prjemian** [comment]
RIght: `xs` is the _prefix_ we define for use in the `nxdl.xsd` file.  The choice of `xs` follows a convention.

clarify: The `http://www.w3.org/2001/XMLSchema` namespace is a URI that just happens to also be a URL (a page available from a web server).

Right: `XML Schema` is one of several (perhaps ~6 in common use?).  They have different strengths.  I believe we _experimented_ with [`schematron`](https://schematron.com/) as an alternative to better express differences between rules for base classes and application definitions.

Right: If we publish our XML Schema as a URL, we could publish both files (`nxdl.xsd` and `nxdlTypes.xsd`).

**Comment by woutdenolf** [comment]
So for XML validation to work for NXDL files located anywhere we would have the rely on `schemaLocation` being a  URL which means

1. In all XML files replaces this

```xml
"xsi:schemaLocation="http://definition.nexusformat.org/nxdl/3.1 ../nxdl.xsd"
```

by this

```xml
xsi:schemaLocation="http://definition.nexusformat.org/nxdl/3.1 http://definition.nexusformat.org/nxdl/nxdl.xsd"
```

2. In `nxdl.xsd` replace this

```xml
<xs:include schemaLocation="nxdlTypes.xsd"> 
```

by this 

```xml
<xs:include schemaLocation="http://definition.nexusformat.org/nxdl/nxdlTypes.xsd"> 
```

3. Deploy `nxdlTypes.xsd` to `http://definition.nexusformat.org/nxdl/nxdlTypes.xsd`

How does the content of  `http://definition.nexusformat.org/nxdl/` get build and deployed? Perhaps URI to URL substitution can be done in that process? We obviously don't want to do it in the .xsd and .xml files in this repository.

**Comment by woutdenolf** [comment]
Additional remarks:

* http://definition.nexusformat.org/nxdl/ contains deployments of the Nexus standard.
* The version of the Nexus Definition Language (i.e. the .xsd files) is the same as the version of the Nexus standard (i.e. the nxdl.xml files)
* http://definition.nexusformat.org/nxdl/ only contains version 3.1 (we are current at v2020.10)

So there are some inconsistencies:

The http://definition.nexusformat.org/nxdl/ structure should be
http://definition.nexusformat.org/nxdl/{version}/nxdl.xsd
http://definition.nexusformat.org/nxdl/{version}/nxdlTypes.xsd
http://definition.nexusformat.org/nxdl/{version}/definitions/*.nxdl.xml

**Comment by woutdenolf** [comment]
@FreddieAkeroyd Could you comment on how to deploy to http://definition.nexusformat.org/nxdl/ from github actions (if possible)?

**Comment by prjemian** [comment]
No work on this issue at 2022-06 Code Camp. Is it necessary to resolve this for release of NXDL now?

**Comment by woutdenolf** [comment]
We can keep it for the next release
